Update TemporaryFolder.newFolder(String) to support passing in paths

kcooney opened the pull request:
that include path separator characters. The overload of newFolder() that
supports passing in multiple strings still does not allow path separators.

marcphilipp commented on src/main/java/org/junit/rules/TemporaryFolder.java:
```diff
@@ -167,16 +169,22 @@ public File newFile() throws IOException {
+        if (folder.startsWith(File.separator)) {
+            throw new IOException("Folder name cannot start with a file separator");
+        }
+        return newFolder(folder.split(FILE_SEPARATOR_REGEX));
```
Instead of the regex, can we use `new File(folder)` and call `getParent()` until it returns `null`? This way a test could use `newFolder("foo/bar")` and be able to run on Windows and Unix.

kcooney replied:
I didn't realize that File supported / as a path separator for Windows.

Instead of your solution, how about I rewite this method to use mkdirs()?

marcphilipp replied:
How would you report already existing intermediate folders (see `newFolder(String[])`) that way?

kcooney replied:
`newFolder(String...)` does not report already existing intermediate directories

marcphilipp replied:
From the code it looks like it does… 😯

marcphilipp replied:
Never mind...

kcooney replied:
Changed to use mkdirs() and added a test for paths that contain a forward slash and for paths that contain the file separator.

marcphilipp commented on src/test/java/org/junit/rules/TemporaryFolderUsageTest.java:
```diff
@@ -84,12 +84,22 @@ public void newFolderWithGivenFolderThrowsIllegalArgumentExceptionIfFolderExists
         tempFolder.newFolder("temp1/temp2");
+        File temp1 = new File(tempFolder.getRoot(), "temp1");
+        assertFileIsDirectory(temp1);
+        assertFileIsDirectory(new File(temp1, "temp2"));
```
This test won't run on Windows, right?

kcooney replied:
Fixed

kcooney commented:
BTW, the behavior changes in 4.12 broke some of our tests. When I am back in the office tomorrow I will verify that this change resolves the problems.

kcooney commented:
PTAL. Thanks, @marcphilipp !

kcooney commented:
I verified that this is a regression from the behavior in JUnit 4.11 and changed the commit comment accordingly.

I also removed the use of Pattern (I forgot to remove that when I changed the code to use `mkdirs`).

Any objections to merging this?

Note I'm also considering changing newFolder(String...) to support paths with slashes, since the inconsistency is weird and supporting slashes would be consistent with `java.nio.file.Paths.get(String, String...)`

marcphilipp commented on src/main/java/org/junit/rules/TemporaryFolder.java:
```diff
@@ -167,16 +167,27 @@ public File newFile() throws IOException {
+     * Returns a new fresh folder with the given path under the temporary
      * folder.
      */
     public File newFolder(String folder) throws IOException {
```
We should probably rename `folder` to `path`, right?

marcphilipp commented on src/main/java/org/junit/rules/TemporaryFolder.java:
```diff
@@ -167,16 +167,27 @@ public File newFile() throws IOException {
+        File file = new File(getRoot(), folder);
+        if (!file.mkdirs()) {
+            throw new IOException(
+                    "a folder with the name \'" + folder + "\' already exists");
```
Is this error message really accurate? I know it's the same as in the other method but I'm sure there are other reasons why a folder could not be created.

marcphilipp commented:
> I verified that this is a regression from the behavior in JUnit 4.11 and changed the commit comment accordingly.
> 
> I also removed the use of `Pattern` (I forgot to remove that when I changed the code to use `mkdirs`).

👍

> Any objections to merging this?

None at all.

> Note I'm also considering changing newFolder(String...) to support paths with slashes, since the inconsistency is weird and supporting slashes would be consistent with `java.nio.file.Paths.get(String, String...)`

I think supporting slashes is great! I've tried to find out why it was implemented that way. From #946 it seems 4.10 used only `mkdir`, 4.11 introduced `newFolder(String...)` but kept using `mkdir`. Supporting slashes should be okay because it would only be more tolerant than before and no existing call should break.

kcooney commented:
@marcphilipp I updated the release notes.

Sorry, I somehow didn't see some of your comments. I'll address them soon.

marcphilipp commented:
Thanks! And no worries about those comments. 🙂
